src/st-trace/trace.c

```diff
@@ -0,0 +1,610 @@
+    // Only check configuration one time after the first 10 seconds of running.
+    uint32_t elapsed_time_s = time(NULL) - trace->start_time;
+    if (trace->configuration_checked || elapsed_time_s < 10) {
+        return;
+    }
+    trace->configuration_checked = true;
+
+    // Simple huristic to determine if we are configured poorly.
+    bool error_no_data = (trace->count_raw_bytes < 100);
+    bool error_low_data = (trace->count_time_packets < 10 && trace->count_target_data < 1000);
+    bool error_bad_data = (trace->count_error > 1 || trace->unknown_sources > 0);
+    bool error_dropped_data = (trace->count_sw_overflow > 0);
+
+    if (!error_no_data && !error_low_data && !error_bad_data && !error_dropped_data)
+        return;
+
+    WLOG("****\n");
+    WLOG("We do not appear to be retrieving data from the stlink correctly.\n");
+
+    if (error_dropped_data) {
+        WLOG("Try setting a slower trace frequency with the --trace=%d command line option.\n", trace_frequency / 2);
+    }
+
+    if (error_no_data || error_low_data || error_bad_data) {
+        uint32_t prescaler;
+        stlink_read_debug32(stlink, TPI_ACPR, &prescaler);
+        if (prescaler) {
+            uint32_t system_clock_speed = (prescaler + 1) * trace_frequency;
+            uint32_t system_clock_speed_mhz = (system_clock_speed + 500000) / 1000000;
+            WLOG("Verify the system clock is running at %d MHz.\n", system_clock_speed_mhz);
+        }
+        WLOG("Try specifying the system clock with the --clock=XX command line option.\n");
+        WLOG("Try setting the trace speed in your device's clock initialization routine:\n");
+        WLOG("  TPI->ACPR = HAL_RCC_GetHCLKFreq() / %d - 1;\n", trace_frequency);
+    }
+
+    WLOG("Diagnostic Information:\n");
+    WLOG("Raw Bytes: %d\n", trace->count_raw_bytes);
+    WLOG("Target Data: %d\n", trace->count_target_data);
+    WLOG("Time Packets: %d\n", trace->count_time_packets);
+    WLOG("Hardware Overflow Count: %d\n", trace->count_hw_overflow);
+    WLOG("Software Overflow Count: %d\n", trace->count_sw_overflow);
+    WLOG("Errors: %d\n", trace->count_error);
+
+    char buffer[1024];
+    memset(buffer, 0, sizeof(buffer));
+    uint32_t offset = 0;
+    for (uint32_t i = 0; i <= 0xFF; i++)
+        if (trace->unknown_opcodes[i / 8] & (1 << i % 8))
+            offset += snprintf(buffer + offset, sizeof(buffer) - offset, "%02x, ", i);
+    WLOG("Unknown Opcodes: %s\n", buffer);
+
+    memset(buffer, 0, sizeof(buffer));
+    offset = 0;
+    for (uint32_t i = 0; i < 32; i++)
+        if (trace->unknown_sources & (1 << i))
+            offset += snprintf(buffer + offset, sizeof(buffer) - offset, "%d, ", i);
+    WLOG("Unknown Sources: %s\n", buffer);
+
+    WLOG("Chip ID: 0x%04x\n", stlink->chip_id);
+    WLOG("****\n");
+}
+
+int main(int argc, char** argv)
+{
+    signal(SIGINT, &abort_trace);
+    signal(SIGTERM, &abort_trace);
+    signal(SIGSEGV, &abort_trace);
+    signal(SIGPIPE, &abort_trace);
```

Comment:
The signal function doesn't work properly on windows, and even more, SIGPIPE didn't define there so this code can't be compiled for windows.
So, if it not hard for you,  can you change the signal processing in a cross-platform way?
Example : https://github.com/stlink-org/stlink/blob/develop/src/st-util/gdb-server.c#L248